Math/NumberTheory/EisensteinIntegers.hs

```diff
@@ -0,0 +1,421 @@
+
+-- | Eisenstein integer division, truncating toward zero.
+quotE :: EisensteinInteger -> EisensteinInteger -> EisensteinInteger
+n `quotE` d = q where (q,_) = quotRemE n d
+
+-- | Eisenstein integer remainder, satisfying
+--
+-- > (x `quotE` y)*y + (x `remE` y) == x
+remE :: EisensteinInteger -> EisensteinInteger -> EisensteinInteger
+n `remE`  d = r where (_,r) = quotRemE n d
+
+-- | Simultaneous 'div' and 'mod' of Eisenstein integers.
+divModE
+    :: EisensteinInteger
+    -> EisensteinInteger
+    -> (EisensteinInteger, EisensteinInteger)
+divModE = divHelper div
+
+-- | Eisenstein integer division, truncating toward negative infinity.
+divE :: EisensteinInteger -> EisensteinInteger -> EisensteinInteger
+n `divE` d = q where (q,_) = divModE n d
+
+-- | Eisenstein integer remainder, satisfying
+--
+-- > (x `divE` y)*y + (x `modE` y) == x
+modE :: EisensteinInteger -> EisensteinInteger -> EisensteinInteger
+n `modE` d = r where (_,r) = divModE n d
+
+-- | Function that does most of the underlying work for @divMod@ and
+-- @quotRem@, apart from choosing the specific integer division algorithm.
+-- This is instead done by the calling function (either @divMod@ which uses
+-- @div@, or @quotRem@, which uses @quot@.)
+divHelper
+    :: (Integer -> Integer -> Integer)
+    -> EisensteinInteger
+    -> EisensteinInteger
+    -> (EisensteinInteger, EisensteinInteger)
+divHelper divide g h =
+    let nr :+ ni = g * conjugate h
+        denom = norm h
+        q = divide nr denom :+ divide ni denom
+        p = h * q
+    in (q, g - p)
+
+-- | Conjugate a Eisenstein integer.
+conjugate :: EisensteinInteger -> EisensteinInteger
+conjugate (a :+ b) = (a - b) :+ (-b)
+
+-- | The square of the magnitude of a Eisenstein integer.
+norm :: EisensteinInteger -> Integer
+norm (a :+ b) = a*a - a * b + b*b
+
+-- | Checks if a given @EisensteinInteger@ is prime. @EisensteinInteger@s
+-- whose norm is a prime congruent to @0@ or @1@ modulo 3 are prime.
+-- See <http://thekeep.eiu.edu/theses/2467 Bandara, Sarada, "An Exposition of the Eisenstein Integers" (2016)>,
+-- page 12.
+isPrime :: EisensteinInteger -> Bool
+isPrime e | e == 0                     = False
+          -- Special case, @1 - ω@ is the only Eisenstein prime with norm @3@,
+          --  and @abs (1 - ω) = 2 + ω@.
+          | a' == 2 && b' == 1         = True
+          | b' == 0 && a' `mod` 3 == 2 = Testing.isPrime a'
+          | nE `mod` 3 == 1            = Testing.isPrime nE
+          | otherwise = False
+  where nE       = norm e
+        a' :+ b' = abs e
+
+-- | Remove @1 - ω@ factors from an @EisensteinInteger@, and calculate that
+-- prime's multiplicity in the number's factorisation.
+divideByThree :: EisensteinInteger -> (Int, EisensteinInteger)
+divideByThree e = go 0 e
+  where
+    go :: Int -> EisensteinInteger -> (Int, EisensteinInteger)
+    go !n !z | r == 0    = go (n + 1) q
+             | otherwise = (n, abs z)
+      where
+        (q, r) = divModE (z * (2 :+ 1)) 3
```

Comment:
Can we use `quotEvenI smth 3` instead of `divModE`? It may be worse to inline and reduce multiplication by `2 :+ 1` to avoid any multiplications: `(a :+ b) * (2 :+ 1) = (a + a - b) :+ (a + b)`.